space/authz/authz.go

```diff
@@ -0,0 +1,191 @@
+	ResourceSetName *string `json:"resource_set_name"`
+	ResourceSetID   *string `json:"resource_set_id"`
+}
+
+// AuthzService represents a space authorization service
+type AuthzService interface {
+	Authorize(ctx context.Context, entitlementEndpoint string, spaceID string) (*string, bool, error)
+	Configuration() AuthzConfiguration
+}
+
+// AuthzConfiguration represents a Keycloak entitlement endpoint configuration
+type AuthzConfiguration interface {
+	GetKeycloakEndpointEntitlement(*goa.RequestData) (string, error)
+}
+
+// AuthzServiceManager represents a space autharizarion service
+type AuthzServiceManager interface {
+	AuthzService() AuthzService
+	EntitlementEndpoint() string
+}
+
+// KeyclaokAuthzServiceManager is a keyaloak implementation of a space autharizarion service
+type KeyclaokAuthzServiceManager struct {
+	Service             AuthzService
+	entitlementEndpoint string
+}
+
+// AuthzService returns a space autharizarion service
+func (m *KeyclaokAuthzServiceManager) AuthzService() AuthzService {
+	return m.Service
+}
+
+// EntitlementEndpoint returns a keyclaok entitlement endpoint URL
+func (m *KeyclaokAuthzServiceManager) EntitlementEndpoint() string {
+	return m.entitlementEndpoint
+}
+
+// KeyclaokAuthzService implements AuthzService interface
+type KeyclaokAuthzService struct {
+	config AuthzConfiguration
+	db     application.DB
+}
+
+// NewAuthzService constructs a new KeyclaokAuthzService
+func NewAuthzService(config AuthzConfiguration, db application.DB) *KeyclaokAuthzService {
+	return &KeyclaokAuthzService{config: config, db: db}
+}
+
+// Configuration returns authz service configuration
+func (s *KeyclaokAuthzService) Configuration() AuthzConfiguration {
+	return s.config
+}
+
+// Authorize returns true and the corresponding Requesting Party Token if the current user is among the space collaborators
+func (s *KeyclaokAuthzService) Authorize(ctx context.Context, entitlementEndpoint string, spaceID string) (*string, bool, error) {
+	token := goajwt.ContextJWT(ctx)
+	if token == nil {
+		return nil, false, errs.NewUnauthorizedError("missing token")
+	}
+	authz := token.Claims.(jwt.MapClaims)["authorization"]
+	if authz == nil {
+		return s.checkEntitlementForSpace(ctx, *token, entitlementEndpoint, spaceID)
+	}
+	var authzJSON authorization
+	err := json.Unmarshal([]byte(authz.(string)), &authzJSON)
+	if err != nil {
+		log.Error(ctx, map[string]interface{}{
+			"space-id": spaceID,
+			"err":      err,
+		}, "unable to unmarshal json with permissions from the rpt token")
+		return nil, false, errs.NewInternalError(fmt.Sprintf("unable to unmarshal json with permissions from the rpt token: %s : %s", authz, err.Error()))
+	}
+	permissions := authzJSON.Permissions
+	if permissions == nil {
+		return s.checkEntitlementForSpace(ctx, *token, entitlementEndpoint, spaceID)
+	}
+	for _, permission := range permissions {
+		name := permission.ResourceSetName
+		if name != nil && spaceID == *name {
+			return &token.Raw, true, nil
```

Comment:
Assuming this is where the UserToken contain a Permissions for a given space etc, shouldn't this still validate if the space resource has changed since the token was issues?